### PR TITLE
quickstart-28: richfaces-validation

### PR DESCRIPTION
**Pull request by lfryc**
based upon quickstart-27: bean-validation

### CODE REVIEW

File: richfaces-validation/README.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+Author: Lukas Fryc <https://community.jboss.org/people/lfryc>
+
+What is it?
+-----------
+
+This is your project! It's a sample, Maven 3 project to help you
+get your foot in the door developing with Java EE 6 on JBoss AS 7 or JBoss Enterprise Application Platform 6. 
+This project is setup to allow you to use JSF 2.0, RichFaces 4.2, CDI 1.0, JPA 2.0 and Bean Validation 1.0. 
+
+It consists from one entity - Member - which is annotated with JSR-303 (Bean Validation)
```
Comment (pmuir) [inline richfaces-validation/README.md]: "consists of" not "consists from"

File: richfaces-validation/README.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+constraints. These constraints are in typical applications checked on several places:
```
Comment (pmuir) [inline richfaces-validation/README.md]: Missing commas "These constraints are, in typical applications, checked on several places:" and also s/on/in/

File: richfaces-validation/README.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+* as database constraints
+* on the persistence layer
+* once entity enters view layer (using JSF / Bean Validation integration)
```
Comment (pmuir) [inline richfaces-validation/README.md]: probably clearer just to say "the view layer"

File: richfaces-validation/README.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+* on the client side (using RichFaces 4.2 - Client Side Validation)
+
+That all in the "define once - check in each layer" manner.
```
Comment (pmuir) [inline richfaces-validation/README.md]: I think you mean to discuss the advantage that this offers - DRY. Perhaps rephrase in these terms?

"This allows you to not repeat the definitions of the constraints, but still have them applied at all layers, and thus report good error messages to users"

File: richfaces-validation/README.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+
+However this sample does not contain any persistence layer, since it shows mainly integration
```
Comment (pmuir) [inline richfaces-validation/README.md]: s/sample/quickstart/

Comment (pmuir) [inline richfaces-validation/README.md]: s/mainly//

File: richfaces-validation/README.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+of RichFaces, JSF and Bean Validation.
+
+There are tests for Bean Validation constraints for Member entity which allows you to check
```
Comment (pmuir) [inline richfaces-validation/README.md]: s/for/for all/

File: richfaces-validation/README.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+implementation without necessity to check the application. These tests are written using Arquillian.
```
Comment (pmuir) [inline richfaces-validation/README.md]: I'm not sure what you mean to say here.

File: richfaces-validation/README.md [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+Application Theme
+-----------------
+
+Application contains view layer written in JSF and RichFaces and shows AJAX-based wizard for new
+member registration. Each member needs to fill in these data in separated screens:
+
+* e-mail
+* name and phone
+* password and its confirmation
```
Comment (pmuir) [inline richfaces-validation/README.md]: Can you remove password, as we don't want to recommend this approach of storing user passwords to our users.

File: richfaces-validation/README.md [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+* confirmation of all inserted data
+
+Once user is successfully registered, he is redirected to initial page with message
```
Comment (pmuir) [inline richfaces-validation/README.md]: s/he is/they are/

Comment (pmuir) [inline richfaces-validation/README.md]: s/with message/with the message/

File: richfaces-validation/README.md [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+that he has been successfully registered.
```
Comment (pmuir) [inline richfaces-validation/README.md]: s/that he has been successfully registered/that they have successfully registered/

File: richfaces-validation/README.md [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+System requirements
+-------------------
+
+All you need to build this project is Java 6.0 (Java SDK 1.6) or better, Maven
+3.0 or better.
+
+Building WAR deployment with Maven
+===============================
+
+To build the application, the only you need is trigger Maven build from
```
Comment (pmuir) [inline richfaces-validation/README.md]: This doesn't really read well in English.

File: richfaces-validation/README.md [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+
+    mvn clean package
+    
+This command will build the WAR archive in target/richfaces-validation.war.
+
+
+Deploying the WAR to the JBoss AS
+=================================
+
+To deploy the application to JBoss AS 7, you needs just copy the built WAR
```
Comment (pmuir) [inline richfaces-validation/README.md:66]: We don't use this approach in the quickstarts, so can you remove it?

File: richfaces-validation/README.md [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+    http://localhost:8080/richfaces-validation/
+
+You can un-deploy the application by deleting the WAR from deployments
+directory or you can use the JBoss AS Maven plugin again:
+
+    mvn jboss-as:undeploy
+    
+In this case, you should see following output in server console:
+
+    JBAS018558: Undeployed "richfaces-validation.war"
```
Comment (pmuir) [inline richfaces-validation/README.md]: I would suggest replacing this with the standard text.

File: richfaces-validation/README.md [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+
+Running the Arquillian tests
+============================
+
+By default, tests are configured to be skipped. The reason is that the sample
+test is an Arquillian test, which requires the use of a container. You can select either
+managed or remote container, the difference is that the remote one requires a running JBoss AS 7 / 
+JBoss Enterprise Application Platform 6 instance prior executing tests.
+
+**Testing on Remote Server**
```
Comment (pmuir) [inline richfaces-validation/README.md]: Can you use markdown syntax here?

File: richfaces-validation/pom.xml [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+        
+        <!-- Import the JSF API, we use provided scope as the API is included 
+         in JBoss AS 7 -->
+        <dependency>
+            <groupId>org.jboss.spec.javax.faces</groupId>
+            <artifactId>jboss-jsf-api_2.0_spec</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        
+        <!-- Import RichFaces runtime dependencies -->
```
Comment (pmuir) [inline richfaces-validation/pom.xml]: Might be worth mentioning that you do include this as a library in the war.

File: richfaces-validation/pom.xml [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+                            <skip>true</skip>
+                        </configuration>
+                    </plugin>
+                    <!-- The JBoss AS plugin deploys your war to a local JBoss
+                        AS container -->
+                    <!-- To use, set the JBOSS_HOME environment variable and run:
+                        mvn package jboss-as:deploy -->
+                    <plugin>
+                        <groupId>org.jboss.as.plugins</groupId>
+                        <artifactId>jboss-as-maven-plugin</artifactId>
```
Comment (pmuir) [inline richfaces-validation/pom.xml]: Shouldn't this be in <plugin> in the main build?

File: richfaces-validation/src/main/java/org/jboss/as/quickstarts/kitchensink/web/RegistrationBean.java [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @Inject
+    Logger logger;
+
+    @Inject
+    transient Flash flash;
+
+    @Inject
+    transient FacesContext facesContext;
```
Comment (pmuir) [inline richfaces-validation/src/main/java/org/jboss/as/quickstarts/kitchensink/web/RegistrationBean.java]: This won't work, as CDI won't reinject on reserialization. Need to find a better way to do this (e.g. inject Instance)

**Comment by pmuir** [comment]
Lukas did you manage to address my comments?

**Comment by lfryc** [comment]
Pete, I have run into issues having unexpected conference engagement.
What is next planned released I shoul target?

**Comment by lfryc** [comment]
Hi Pete, I have been able to incorporate all comments in README and removed password from registration wizard.

**Comment by sgilda** [comment]
Hi Lukas,
I'm testing your quickstart for usability. It builds and deploys great, however, I ran into a couple of problems and have a few comments.

-- Could you add the command line to deploy to the quickstart README? I'm usually in a hurry and don't use an IDE and I'm guessing there are others out there like me. ;-)
           mvn jboss-as:deploy
-- Could you change the web context from richfaces-validation to jboss-as-richfaces-validation to be consistent with the other quickstarts? It also assures this one will be unique.
-- I was playing around with validation. 
   ---- When I entered 'sgilda' and clicked on "Continue", it gave an error "not an well-formed e-mail". So far, so good.
   ---- Then I entered 'sgilda@red' and clicked on "Continue". It cleared up the errors, but stayed on the same screen. Shouldn't it go to the next screen?
   ---- I clicked continue once more and it this time it took me to the next screen.
   ---- On the Registration form page, I decided to click on "Back" to go back and append the ".com" to my email address. It gave me validation errors for both fields. It shows Name: "size must be between 1 and 25" and "Phone number must be 10-12 digits". Should it be performing validation when I click the "Back" button?

Thanks, Sande

**Comment by lfryc** [comment]
Hello Sande, I have addressed all the issues you mentioned, except the last one:

"---- On the Registration form page, I decided to click on "Back" to go back and append the ".com" to my email address. It gave me validation errors for both fields. It shows Name: "size must be between 1 and 25" and "Phone number must be 10-12 digits". Should it be performing validation when I click the "Back" button?"

Ideally it should work so, but we have run there into issue with the RichFaces <rich:toggleControl /> component.
It will be addressed in upcoming releases:

https://issues.jboss.org/browse/RF-12068

Since there is no simple workaround,
we have two options: rework it to not use this component
or accept it as it is.

**Comment by sgilda** [comment]
Well, it's ugly, but not worth the effort since it's a bug in the underlying component. 

Pete, do you agree? 

I'll check out the changes.

**Comment by sgilda** [comment]
Looks great! I have a few minor nit-picking comments. 
In the README, could you change the following:
Change the title line to read: richfaces-validation: RichFaces and Bean Validation
Change the access URL to a link so when they click on it, it takes them to your running application
    http://localhost:8080/jboss-as-richfaces-validation/

Other than that, this is good to go. Thanks!

Pete, do you need to review anything?

**Comment by pmuir** [comment]
Made updates and merged.

**Comment by lfryc** [comment]
Great, thanks Pete, Sande.

**Pull request closed without merging**